examples/server/server.cpp

```diff
@@ -2772,6 +2783,38 @@ int main(int argc, char **argv)
         res.set_header("Access-Control-Allow-Headers", "*");
     });
 
+    // Middleware for API key validation
+    auto validate_key = [&sparams](const httplib::Request &req, httplib::Response &res, std::vector<std::string> &keys) -> bool {
+        // If API key is not set, skip validation
+        if (keys.empty()) {
+            return true;
+        }
+
+        // Check for API key in the header
+        auto auth_header = req.get_header_value("Authorization");
+        std::string prefix = "Bearer ";
+        if (auth_header.substr(0, prefix.size()) == prefix) {
+            std::string received_api_key = auth_header.substr(prefix.size());
+            if (std::find(keys.begin(), keys.end(), received_api_key) != keys.end()) {
+                return true; // API key is valid
+            }
+        }
+
+        // Check for API key in the params
+        auto auth_param = req.get_param_value("key");
```

Comment:
Query params are in clear on all http traffic scanners. I feel this hack is a security breach.
Monitoring endpoints must simply not be exposed to the world, or at least slots data simply can be disabled. They are here for debug purpose during development only.
@ggerganov I would prefer we invest on a real security protocol like Oauth2/openid than adding this.

Reply:
Right, but authorization headers are also passed in cleartext without https. Given the nature of the api keys as they are at the moment this is no more or less secure than the current solution.

Long term, specifying keys on the command line and passing them directly is not the correct solution.

This is just a quick way to get some level of security on those endpoints in the server as it exists now until a proper (jwt would make sense to me) solution is in place.

Reply:
I am speaking about logging. Authorization header is never logged. Query params are, example:

https://github.com/ggerganov/llama.cpp/blob/9731134296af3a6839cd682e51d9c2109a871de5/examples/server/server.cpp#L2691-L2705

But all http monitoring tool will do the same.

Note: JWT is not a security protocol, just a transparent token often used in oauth2.